File: c/src/bluez/agent.c
#include <agent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Introspection XML for org.bluez.Agent1 */
static const gchar introspection_xml[] =
    "<node>"
    "  <interface name='org.bluez.Agent1'>"
    "    <method name='Release'/>"
    "    <method name='RequestPasskey'>"
    "      <arg type='o' name='device' direction='in'/>"
    "      <arg type='u' name='passkey' direction='out'/>"
    "    </method>"
    "    <method name='DisplayPasskey'>"
    "      <arg type='o' name='device' direction='in'/>"
    "      <arg type='u' name='passkey' direction='in'/>"
    "    </method>"
    "    <method name='RequestConfirmation'>"
    "      <arg type='o' name='device' direction='in'/>"
    "      <arg type='u' name='passkey' direction='in'/>"
    "    </method>"
    "    <method name='RequestPinCode'>"
    "      <arg type='o' name='device' direction='in'/>"
    "      <arg type='s' name='pincode' direction='out'/>"
    "    </method>"
    "    <method name='RequestAuthorization'>"
    "      <arg type='o' name='device' direction='in'/>"
    "    </method>"
    "    <method name='AuthorizeService'>"
    "      <arg type='o' name='device' direction='in'/>"
    "      <arg type='s' name='uuid' direction='in'/>"
    "    </method>"
    "  </interface>"
    "</node>";

/* Structure definition, private to agent.c */
struct _Agent
{
    GDBusConnection *connection;
    gchar *object_path;
    guint registration_id;
    GDBusNodeInfo *node_info;
};

/* Method call handler for the Agent interface */
static void
handle_method_call(GDBusConnection *connection,
                   const gchar *sender,
                   const gchar *object_path,
                   const gchar *interface_name,
                   const gchar *method_name,
                   GVariant *parameters,
                   GDBusMethodInvocation *invocation,
                   gpointer user_data)
{
    if (g_strcmp0(method_name, "Release") == 0)
    {
        g_print("Agent Released\n");
        g_dbus_method_invocation_return_value(invocation, NULL);
    }
    else if (g_strcmp0(method_name, "RequestPasskey") == 0)
    {
        const gchar *device;
        g_variant_get(parameters, "(&o)", &device);
        g_print("RequestPasskey for device: %s\n", device);
        g_dbus_method_invocation_return_value(invocation, g_variant_new("(u)", 0));
    }
    else if (g_strcmp0(method_name, "DisplayPasskey") == 0)
    {
        const gchar *device;
        guint passkey;
        g_variant_get(parameters, "(&ou)", &device, &passkey);
        g_print("DisplayPasskey for device: %s, passkey: %u\n", device, passkey);
        g_dbus_method_invocation_return_value(invocation, NULL);
    }
    else if (g_strcmp0(method_name, "RequestConfirmation") == 0)
    {
        const gchar *device;
        guint passkey;
        g_variant_get(parameters, "(&ou)", &device, &passkey);
        g_print("Auto-confirming for device: %s, passkey: %u\n", device, passkey);
        g_dbus_method_invocation_return_value(invocation, NULL);
    }
    else if (g_strcmp0(method_name, "RequestPinCode") == 0)
    {
        const gchar *device;
        g_variant_get(parameters, "(&o)", &device);
        g_print("RequestPinCode for device: %s\n", device);
        g_dbus_method_invocation_return_value(invocation, g_variant_new("(s)", "0000"));
    }
    else if (g_strcmp0(method_name, "RequestAuthorization") == 0)
    {
        const gchar *device;
        g_variant_get(parameters, "(&o)", &device);
        g_print("RequestAuthorization for device: %s\n", device);
        g_dbus_method_invocation_return_value(invocation, NULL);
    }
    else if (g_strcmp0(method_name, "AuthorizeService") == 0)
    {
        const gchar *device;
        const gchar *uuid;
        g_variant_get(parameters, "(&os)", &device, &uuid);
        g_print("AuthorizeService for device: %s with uuid: %s\n", device, uuid);
        g_dbus_method_invocation_return_value(invocation, NULL);
    }
    else
    {
        g_print("Unknown method: %s\n", method_name);
        g_dbus_method_invocation_return_error(invocation,
                                              G_IO_ERROR,
                                              G_IO_ERROR_INVALID_ARGUMENT,
                                              "Unknown method: %s", method_name);
    }
}

/* vtable for the Agent interface */
static const GDBusInterfaceVTable agent_vtable = {
    handle_method_call,
    NULL,
    NULL};

/* Create a new Agent instance */
Agent *agent_new(GDBusConnection *connection, const char *object_path)
{
    Agent *agent = malloc(sizeof(Agent));
    if (!agent)
        return NULL;
    agent->connection = connection;
    agent->object_path = g_strdup(object_path);
    agent->node_info = g_dbus_node_info_new_for_xml(introspection_xml, NULL);
    if (!agent->node_info)
    {
        g_free(agent->object_path);
        free(agent);
        return NULL;
    }
    return agent;
}

/* Register the Agent on the D-Bus connection */
void agent_register(Agent *agent)
{
    GError *error = NULL;
    agent->registration_id = g_dbus_connection_register_object(
        agent->connection,
        agent->object_path,
        agent->node_info->interfaces[0],
        &agent_vtable,
        agent,
        NULL,
        &error);
    if (agent->registration_id == 0)
    {
        g_printerr("Failed to register agent at %s: %s\n", agent->object_path, error->message);
        g_error_free(error);
    }
    else
    {
        g_print("Agent registered at %s\n", agent->object_path);
    }
}

/* Free Agent memory and unregister the object */
void agent_free(Agent *agent)
{
    if (!agent)
        return;
    if (agent->registration_id > 0)
        g_dbus_connection_unregister_object(agent->connection, agent->registration_id);
    if (agent->node_info)
        g_dbus_node_info_unref(agent->node_info);
    g_free(agent->object_path);
    free(agent);
}

File: rust/src/bluez/agent.rs
use std::error::Error;
use zbus::{fdo, interface, Connection, Proxy};
use zbus::zvariant::{ObjectPath, Value};
use crate::constants::{ADAPTER_IFACE, ADAPTER_PATH, AGENT_MANAGER_IFACE, BLUEZ_SERVICE, BLUEZ_SERVICE_PATH, DBUS_PROPERTIES_IFACE};
use crate::utils::ObjectPathTrait;

#[derive(Default, Debug)]
pub struct Agent {
    pub path: String,
}

impl Agent {
    pub fn new(path: String) -> Self {
        Self { path }
    }
}

impl ObjectPathTrait for Agent {
    fn object_path(&self) -> String {
        self.path.clone()
    }
}

#[interface(name = "org.bluez.Agent1")]
impl Agent {
    // The Release method: no parameters, no return.
    async fn release(&self) -> fdo::Result<()> {
        println!("Agent Released");
        Ok(())
    }

    // The RequestPasskey method receives a device object path and returns an u32.
    async fn request_passkey(&self, device: ObjectPath<'_>) -> fdo::Result<u32> {
        println!("RequestPasskey for device: {}", device);
        Ok(0)
    }

    // The DisplayPasskey method receives a device path and a passkey.
    async fn display_passkey(&self, device: ObjectPath<'_>, passkey: u32) -> fdo::Result<()> {
        println!(
            "DisplayPasskey for device: {}, passkey: {}",
            device, passkey
        );
        Ok(())
    }

    // The RequestConfirmation method receives a device path and passkey.
    async fn request_confirmation(&self, device: ObjectPath<'_>, passkey: u32) -> fdo::Result<()> {
        println!(
            "Auto-confirming pairing for device: {} with passkey: {}",
            device, passkey
        );
        Ok(())
    }

    // The RequestPinCode method receives a device path and returns a string.
    async fn request_pin_code(&self, device: ObjectPath<'_>) -> fdo::Result<String> {
        println!("RequestPinCode for device: {}", device);
        Ok("0000".into())
    }

    // The RequestAuthorization method receives a device path.
    async fn request_authorization(&self, device: ObjectPath<'_>) -> fdo::Result<()> {
        println!("RequestAuthorization for device: {}", device);
        Ok(())
    }

    // The AuthorizeService method receives a device path and a service UUID.
    async fn authorize_service(&self, device: ObjectPath<'_>, uuid: &str) -> fdo::Result<()> {
        println!(
            "AuthorizeService called for device: {} and service UUID: {}",
            device, uuid
        );
        Ok(())
    }
}

pub async fn register_agent(
    connection: &Connection,
    agent_object_path: &str,
    capability: &str,
) -> Result<(), zbus::Error> {
    // Create a proxy for the AgentManager interface on /org/bluez.
    let agent_manager = Proxy::new(
        &connection,
        BLUEZ_SERVICE,
        BLUEZ_SERVICE_PATH,
        AGENT_MANAGER_IFACE,
    )
    .await?;

    println!("Agent manager proxy created");

    let result_registering = agent_manager
        .call_method(
            "RegisterAgent",
            &(ObjectPath::try_from(agent_object_path)?, capability),
        )
        .await?;

    println!("Agent registered: {:?}", result_registering);

    // Call RequestDefaultAgent(agent_object_path)
    agent_manager
        .call_method(
            "RequestDefaultAgent",
            &(ObjectPath::try_from(agent_object_path)?),
        )
        .await?;
    println!("Agent registered as default with {} capability", capability);

    // Attempt to set adapter properties using the Properties interface.
    match Proxy::new(
        &connection,
        BLUEZ_SERVICE,
        ADAPTER_PATH,
        DBUS_PROPERTIES_IFACE,
    )
    .await
    {
        Ok(adapter_proxy) => {
            async fn set_property(
                proxy: &Proxy<'_>,
                property: &str,
                value: bool,
            ) -> Result<(), Box<dyn Error>> {
                proxy
                    .call_method("Set", &(ADAPTER_IFACE, property, Value::from(value)))
                    .await?;
                Ok(())
            }
            if let Err(e) = set_property(&adapter_proxy, "Powered", true).await {
                eprintln!("Error setting Powered: {}", e);
            }
            if let Err(e) = set_property(&adapter_proxy, "Discoverable", true).await {
                eprintln!("Error setting Discoverable: {}", e);
            }
            if let Err(e) = set_property(&adapter_proxy, "Pairable", true).await {
                eprintln!("Error setting Pairable: {}", e);
            }
        }
        Err(e) => {
            eprintln!("Error obtaining adapter proxy: {}", e);
        }
    }

    Ok(())
}
